### PR TITLE
Adding automatic scoring based on standard bingo templates.

### PR DESCRIPTION
**Pull request by martiansideofthemoon**

### CODE REVIEW

File: bingo.py [hunk 1]
```diff
@@ -50,7 +58,10 @@ def print(self):
         assert len(o) == n
 
         random.shuffle(o)
+        self.linear_grid = o
 
+    def print(self):
```
Comment (mrdrozdov) [inline bingo.py:63]: Thanks these should be separated.

File: bingo.py [hunk 2]
```diff
@@ -70,8 +81,19 @@ def print(self):
 
         print('')
 
-
-
+    def score(self):
+        free_ids = [x[0] * self.w + x[1] for x in self.free]
```
Comment (mrdrozdov) [inline bingo.py:85]: I see. This converts the coordinate into an array index.

File: bingo.py [hunk 3]
```diff
@@ -70,8 +81,19 @@ def print(self):
...
+        checked_str = [
+            1 if self.linear_grid[i] in self.chosen_o or i in free_ids else 0
+            for i in range(self.h * self.w)
+        ]
```
Comment (mrdrozdov) [inline bingo.py:89]: Could have done:
```
checked_str = [1 if self.linear_grid[i] in self.chosen_o or 0 for i in range(self.h * self.w)]
for i in free_ids:
  checked_str[i] = 1
```

But what you did works too.

File: bingo.py [hunk 4]
```diff
@@ -103,7 +125,13 @@ def main():
         board = Board(global_o)
         board.read(path)
         board.add(chosen_o)
+        board.make_grid()
         board.print()
+        found = board.score()
+        if len(found) > 0:
+            print("Score = {:d} (You found {})\n".format(len(found), ", ".join(found)))
+        else:
+            print("Score = 0 (You got lost in the maze!)\n")
```
Comment (mrdrozdov) [inline bingo.py:134]: 😻

File: bingo.py [hunk 5]
```diff
@@ -70,8 +81,19 @@ def print(self):
...
+        checked_str = [
+            1 if self.linear_grid[i] in self.chosen_o or i in free_ids else 0
+            for i in range(self.h * self.w)
+        ]
+        found = []
+        for temp_name, template in TEMPLATES.items():
+            temp_list = temp2list(template)
+            if check(checked_str, temp_list):
+                found.append(temp_name)
+        found.sort()
```
Comment (mrdrozdov) [inline bingo.py:95]: If you sort ahead of time by name, then don't need this call.

But my comments are for easy readability, and your code has been written for optimization 🙃

**Comment by mrdrozdov** [comment]
LGTM! :shipit:

File: bingo.py (hunk 5 above)
Comment (martiansideofthemoon) [reply bingo.py:95]: Sorry, didn't get this part. The main reason I have a sort is to fix the order of displayed matches. An ordered dictionary for templates will also do the trick and avoid the sorting.

**Pull request merged**